composeml/label_maker.py

```diff
@@ -4,146 +4,376 @@
-                return labels
+                continue
+
+            n_examples = 0
+
+            for cutoff_time, gap_end in intervals:
+                df_slice, window_end = self.get_slice(df=df, cutoff_time=cutoff_time, gap_end=gap_end)
+
+                if df_slice.empty:
+                    continue
+
+                n_examples += 1
+
+                if verbose:
+                    info = {
+                        self.target_entity: key,
+                        'slice': n_examples,
+                        'window': f'[{cutoff_time}, {window_end})',
+                        'gap': f'[{cutoff_time}, {gap_end})',
+                    }
+                    info = pd.Series(info).to_string()
+                    print(info, end='\n\n')
 
-            cutoff_time = offset_time(df.index, minimum_data)
+                if edges:
+                    df_slice = df_slice, (cutoff_time, window_end)
 
-            for example in range(num_examples_per_instance):
-                df = df[cutoff_time:]
+                yield df_slice
 
-                if df.empty:
-                    skipped_iterations = num_examples_per_instance - example
-                    progress_bar.update(n=skipped_iterations)
+                if n_examples >= num_examples_per_instance:
                     break
 
-                window_end = offset_time(df.index, self.window_size)
-                label = self.labeling_function(df[:window_end], *args, **kwargs)
+    def search(self, df, num_examples_per_instance, minimum_data=None, gap=None, verbose=True, *args, **kwargs):
+        """Searches the data to calculates labels.
+
+        Args:
+            df (DataFrame) : Data frame to search and extract labels.
+            num_examples_per_instance (int) : Number of examples per unique instance of target entity.
+            minimum_data (str) : Minimum data before starting search.
+            gap (str) : Time between examples.
+            verbose (bool) : Whether to render progress bar.
+            *args : Positional arguments for labeling function.
+            **kwargs : Keyword arguments for labeling function.
+
+        Returns:
+            LabelTimes : Calculated labels with cutoff times.
+        """
+        bar_format = "Elapsed: {elapsed} | Remaining: {remaining} | "
+        bar_format += "Progress: {l_bar}{bar}| "
+        bar_format += self.target_entity + ": {n}/{total} "
+        n_groups = df.groupby(self.target_entity).ngroups
+        total = n_groups * num_examples_per_instance
 
-                if not pd.isnull(label):
-                    labels[cutoff_time] = label
+        if num_examples_per_instance == -1 or num_examples_per_instance == float('inf'):
+            disable = True
+        else:
+            disable = not verbose
 
-                cutoff_time = offset_time(df.index, gap)
-                progress_bar.update(n=1)
+        progress_bar = tqdm(total=total, bar_format=bar_format, disable=disable, file=stdout)
```

Comment:
When user sets the `num_examples_per_instance` to all possible examples, how do we update the progress bar?

Reply:
it should be updated only when an instance is finished

Reply:
updated progress bar